### DISCUSSION TITLE
Provisioning different devices

### DISCUSSION THREAD

**Discussion by hgschmie** [toitlang/toit#1330] [Q&A] [answered]
I have a bunch of devices around the house to measure environment data. They are all the same, except that they have slightly different sensors attached and/or should log their data to different mqtt topics.

I have this working with some old and brittle C code that I want to get rid of. Basically, when I flash the device, I add a per-device custom element that contains the device provisioning. 

My hope was to do a similar thing with toit/jaguar. All devices get the same software and then I would put a provision file on the device giving it its identity, what sensors, what mqtt topics and so on. 

How would I do that? I can have the devices come up, report their name to a service and pull that provision file down, but this is cumbersome and requires infrastructure. Is there a way for me to give a provision file when I flash the device (not ideal but ok) or when I install the software with jaguar? I looked into assets but I wanted to avoid recompiling the firmware.

**Comment by floitsch** [comment]
Note that assets don't require you to recompile the firmware. Adding an asset consists of adding new data to the firmware envelope without changing the elf file that is in the envelope. @kasperl, do we expose the asset functionality nicely in Jaguar?

That said, there is probably an easier way. You could just store the configuration in flash (see https://libs.toit.io/device/class-FlashStore).
You could either `jag run` a program that saves the corresponding information in the flash, or do it like Jaguar: start a small http server that accepts the configuration and writes it to flash.

I could imagine the following work-flow:
- flash Jaguar
- modify a config.toit file
- `jag run config.toit` which then writes the configuration to flash.

Alternatively; more sophisticated:
- flash Jaguar
- `jag run config_updater.toit`  which launches an http server
- push the configuration with `curl`, a separate tool, or using a web-browser (assuming `config_updater` serves a nice html page).

**Comment by kasperl** [comment, marked as answer]
Assets should work well for this case. You can create the firmware.envelope file and flash it onto a device. Once you install a container on the device using `jag container install` or run code there using `jag run` you can add the device specific assets.

You create the assets file using `sdk/tools/assets` like described here: https://github.com/toitlang/toit#adding-container-assets. You can then add the assets when you run like this:

```
jag run --assets=device.assets program.toit
```

From within `program.toit` you should now be able to see the assets by importing `system.assets` and calling `assets.decode`.

**Comment by hgschmie** [comment]
Turned out to be pretty straightforward. No idea how I did not notice the `--assets` option for container and run before.

```
build/host/sdk/tools/assets -e device-1.assets create
build/host/sdk/tools/assets -e device-1.assets add config device-1.json
jag flash --name device-1
jag container install --device device-1 --assets=device-1.assets light-monitor sensor-monitor.toit
```

rinse, repeat for device 2, 3, 4 etc. I generate the json files from a script locally. 

Super-easy. Thank you! I think toit already saved me more time fuzzing around with SDKs, cross compilers and all that pain than I put into the first version of these little critters.
